### PR TITLE
fix: Preserve trailing slash in service invocation via dapr-app-id header

### LINKED ISSUE #7686: Service invocation via request header removes trailing slash
Hi,

this week we tried to implement a Non-Dapr endpoint via HTTP. Normally we use the `dapr-app-id` in the header instead of the `/v1.0/invoke/{appId}/method/{methodName}`, but ran into an issue with one of our endpoints. For this endpoint we need to have a trailing slash, which doesn't arrive at our server. Strangely enough, the dapr logs show the trailing slash. Calling the endpoint with the `/v1.0/invoke/{appId}/method/{methodName}/` version works. So my guess is, that it has something to to with the `dapr-app-id` header version. (Maybe the log is not correct?)

## In what area(s)?

/area runtime

## What version of Dapr?

> 1.13.0

## Expected Behavior

URL should be send to the endpoint as specified.

## Actual Behavior

Trailing slash is missing. 

## Steps to Reproduce the Problem

Dapr Kubernetes Sidecar
Calling the HTTPEndpoint via `/v1.0/invoke/{appId}/method/{methodName}/` (including trailing slash), the server returns 200.

Dapr Log
```
time="2024-04-10T13:47:41.873135982Z" level=info msg="HTTP API Called" app_id=xy code=200 duration=1159 instance=xy-6d5bb584b-s4trc method="GET /v1.0/invoke/appId/method/my-demo/route/2024-04-08T00:00:00.0000000/2024-04-08T23:59:59.9990000/" scope=dapr.runtime.http-info size=4994 type=log ver=1.13.0
```

Server Log
```
2024-04-10 13:47:41 10.21.1.10 GET /my-demo/route/2024-04-08T00:00:00.0000000/2024-04-08T23:59:59.9990000/ - 443 - 127.0.0.1 Go-http-client/2.0 - 200 0 0 1065 400 6610 800d
```

Calling the HTTPEndpoint via `/{methodName}/` and `dapr-app-id` header, the server returns 404 (as the trailing slash is missing)

Dapr Log
```
time="2024-04-10T13:50:03.933427538Z" level=info msg="HTTP API Called" app_id=xy code=404 duration=11 instance=xy-6d5bb584b-s4trc method="GET /my-demo/route/2024-04-08T00:00:00.0000000/2024-04-08T23:59:59.9990000/" scope=dapr.runtime.http-info size=5848 type=log ver=1.13.0
```

Server Log 
```
2024-04-10 13:50:03 10.21.1.10 GET /my-demo/route/2024-04-08T00:00:00.0000000/2024-04-08T23:59:59.9990000 - 443 - 127.0.0.1 Go-http-client/2.0 - 404 0 2 10 400 6610 800d
```

## Release Note

RELEASE NOTE:

### PR DESCRIPTION
**Pull request by franciscocunha55**
## Description

When using the `dapr-app-id` header for service invocation, trailing slashes
are stripped from the request path due to `path.Clean()` in
`findTargetIDAndMethod()`. This causes 404 errors for services that
distinguish `/endpoint` from `/endpoint/`.

The fix preserves trailing slashes after path normalization, making the
behavior consistent with URL-based invocation (`/v1.0/invoke/{appId}/method/{method}/`).

Fixes: #7686

## Changes

- Modified `findTargetIDAndMethod()` in `pkg/api/http/directmessaging.go`
  to preserve trailing slashes after `path.Clean()`
- Added test cases for trailing slash preservation

## Testing

- Added unit tests for trailing slash scenarios (header, basic auth, nested paths)
- All existing tests pass
- `make lint` passes

### CODE REVIEW

File: pkg/api/http/directmessaging_test.go [hunk 1]
```diff
@@ -1074,6 +1074,10 @@ func TestFindTargetIDAndMethod(t *testing.T) {
 		{name: "path with https target escaped", path: "/v1.0/invoke/https%3A%2F%2Fexample.com/method/foo", wantTargetID: "https://example.com", wantMethod: "foo"},
 		{name: "path with https target partly escaped", path: "/v1.0/invoke/https%3A/%2Fexample.com/method/foo", wantTargetID: "https://example.com", wantMethod: "foo"},
 		{name: "extra slashes are removed", path: "///foo//bar", headers: http.Header{"Dapr-App-Id": []string{"myapp"}}, wantTargetID: "myapp", wantMethod: "foo/bar"},
+		{name: "dapr-app-id header preserves trailing slash", path: "/foo/bar/", headers: http.Header{"Dapr-App-Id": []string{"myapp"}}, wantTargetID: "myapp", wantMethod: "foo/bar/"},
+		{name: "dapr-app-id header preserves trailing slash on nested path", path: "/api/v1/resource/", headers: http.Header{"Dapr-App-Id": []string{"myapp"}}, wantTargetID: "myapp", wantMethod: "api/v1/resource/"},
+		{name: "basic auth preserves trailing slash", path: "/foo/bar/", headers: http.Header{"Authorization": []string{"Basic ZGFwci1hcHAtaWQ6YXV0aA=="}}, wantTargetID: "auth", wantMethod: "foo/bar/"},
```
Comment (acroca) [inline pkg/api/http/directmessaging_test.go:1079]: You could add  a test for `/foo/bar///` -> `/foo/bar/`

Comment (franciscocunha55) [reply pkg/api/http/directmessaging_test.go:1079]: done

File: pkg/api/http/directmessaging.go [hunk 2]
```diff
@@ -366,7 +366,11 @@ func (a *api) onDirectMessage(w http.ResponseWriter, r *http.Request) {
 // 3. URL parameter: `http://localhost:3500/v1.0/invoke/<app-id>/method/<method>`
 func findTargetIDAndMethod(reqPath string, headers http.Header) (targetID string, method string) {
 	if appID := headers.Get(consts.DaprAppIDHeader); appID != "" {
-		targetID, method = appID, strings.TrimPrefix(path.Clean(reqPath), "/")
+		cleaned := path.Clean(reqPath)
+		if len(reqPath) > 1 && strings.HasSuffix(reqPath, "/") && cleaned != "/" {
+			cleaned += "/"
+		}
+		targetID, method = appID, strings.TrimPrefix(cleaned, "/")
```
[the remark this answers is not in the document]
Comment (acroca) [reply pkg/api/http/directmessaging.go]: I was going to suggest this same thing, this code is duplicated and I think it's easy to fix.

[the remark this answers is not in the document]
Comment (franciscocunha55) [reply pkg/api/http/directmessaging.go]: Refactored the duplicated logic into a `cleanPathPreserveTrailingSlash()` helper function. Also added tests for multiple trailing slashes and the root-collapsing edge case.

**Review by acroca — APPROVED** [review]
